File: app/onboarding.py
"""First-launch onboarding dialog for macOS.

Shown when ~/.anyclip/config.json is missing. Returns the chosen token
(either freshly generated or pasted from the first device), or None if
the user cancelled. PyObjC NSAlert keeps the dependency surface to
what py2app already bundles -- no Tk, no extra wheels.
"""

from __future__ import annotations

import logging
from typing import Optional

import config_store

log = logging.getLogger("anyclip.onboarding")


def show_onboarding() -> Optional[str]:
    """Return the chosen token or None if cancelled.

    On non-macOS hosts this falls back to generating a token silently
    so the GUI path stays runnable for smoke tests off-platform.
    """
    try:
        from AppKit import (  # type: ignore[import-not-found]
            NSAlert,
            NSMakeRect,
            NSTextField,
        )
    except ImportError:
        log.warning("PyObjC not available; auto-generating token")
        return config_store.generate_token()

    alert = NSAlert.alloc().init()
    alert.setMessageText_("Welcome to AnyClip")
    alert.setInformativeText_(
        "Choose how to set the shared clipboard token. Both devices "
        "must use the same value."
    )
    alert.addButtonWithTitle_("Generate new token (first device)")
    alert.addButtonWithTitle_("Enter existing token (second device)")
    alert.addButtonWithTitle_("Cancel")

    response = alert.runModal()
    # NSAlertFirstButtonReturn = 1000, then 1001, 1002, ...
    if response == 1000:
        return config_store.generate_token()
    if response == 1001:
        return _prompt_for_token(NSAlert, NSTextField, NSMakeRect)
    return None


def _prompt_for_token(NSAlert, NSTextField, NSMakeRect) -> Optional[str]:
    alert = NSAlert.alloc().init()
    alert.setMessageText_("Enter shared token")
    alert.setInformativeText_(
        "Paste the token shown on your first device."
    )
    alert.addButtonWithTitle_("OK")
    alert.addButtonWithTitle_("Cancel")

    field = NSTextField.alloc().initWithFrame_(NSMakeRect(0, 0, 320, 24))
    alert.setAccessoryView_(field)
    response = alert.runModal()
    if response != 1000:
        return None
    value = field.stringValue()
    return str(value).strip() or None

File: formacOS/Sources/AnyClipApp/Onboarding.swift
import AppKit
import AnyClipCore

/// First-launch token dialog. Port of app/onboarding.py + the resolution
/// order from menubar_mac._resolve_token: env > on-disk config > dialog.
@MainActor
enum Onboarding {
    static func resolveToken() -> String? {
        if let env = ProcessInfo.processInfo.environment["ANYCLIP_TOKEN"],
           !env.isEmpty {
            return env
        }
        if let stored = ConfigStore.load() {
            return stored.token
        }
        guard let token = show() else { return nil }
        try? ConfigStore.save(StoredConfig(token: token))
        return token
    }

    private static func show() -> String? {
        let alert = NSAlert()
        alert.messageText = "Welcome to AnyClip"
        alert.informativeText =
            "Choose how to set the shared clipboard token. Both devices "
            + "must use the same value."
        alert.addButton(withTitle: "Generate new token (first device)")
        alert.addButton(withTitle: "Enter existing token (second device)")
        alert.addButton(withTitle: "Cancel")
        switch alert.runModal() {
        case .alertFirstButtonReturn:
            return ConfigStore.generateToken()
        case .alertSecondButtonReturn:
            return promptForToken()
        default:
            return nil
        }
    }

    private static func promptForToken() -> String? {
        let alert = NSAlert()
        alert.messageText = "Enter shared token"
        alert.informativeText = "Paste the token shown on your first device."
        alert.addButton(withTitle: "OK")
        alert.addButton(withTitle: "Cancel")
        let field = NSTextField(frame: NSRect(x: 0, y: 0, width: 320, height: 24))
        alert.accessoryView = field
        alert.window.initialFirstResponder = field
        guard alert.runModal() == .alertFirstButtonReturn else { return nil }
        let value = field.stringValue.trimmingCharacters(in: .whitespacesAndNewlines)
        return value.isEmpty ? nil : value
    }
}
